packages/@aws-cdk/assets/lib/fs/utils.ts

```diff
@@ -58,3 +59,79 @@ export function shouldFollow(mode: FollowMode, sourceRoot: string, realPath: str
     return path.resolve(realPath).startsWith(path.resolve(sourceRoot));
   }
 }
+
+export function listFilesRecursively(dir: string, options: CopyOptions & Required<Pick<CopyOptions, 'follow'>>, rootDir?: string): string[] {
```

Comment:
Why not make it easy on consumers and instead of returning a list of strings, return a list of objects with metadata?

```ts
interface AssetFile { 
  relativePath: string;
  absolutePath: string;
  isSymlink: boolean;
  isDirectory: boolean;
  symlinkTarget?: string;
}

 export function listFilesRecursively(...): AssetFile[];
```

Avoids having consumers have to do all that `stat`ing again.

In fact you could even do:

```ts
type AssetFile = { 
  relativePath: string;
  absolutePath: string;
  isDirectory: boolean;
} & ({ isSymlink: false } | { isSymlink: true; symlinkTarget: string});
```

To benefit from that sweet sweet TypeScript type checking.

Reply:
Fun type 👍 
And done